Realm/RLMSyncUser.h

```diff
@@ -113,8 +140,118 @@ NS_SWIFT_UNAVAILABLE("Use the full version of this API.");
+ passed into `-acceptOffer:callback:` to accept the offer. Do not confuse
+ this token with the identifier returned immediately by this method.
+
+ @param realmURL        The URL of the synced Realm on the server.
+ @param date            The date when the offer should expire. If nil, the
+                        offer will never expire.
+ @param permissions     The permissions that the offer should include.
+ @param token           A pointer to a notification token object, through
+                        which a notification token should be passed out.
+ @param callback        A block through which the status of the permission
+                        offer creation process should be reported.
+ @return    An identifier for the newly-created permission offer. This
+            identifier can be passed into any method taking an `offerKey`
+            parameter. Do not confuse this with the permission offer's
+            token, which is passed back asynchronously through the callback
+            once the server has processed the request.
+ */
+- (NSString *)offerPermissionsForURL:(NSURL *)realmURL
+                          expiration:(nullable NSDate *)date
+                         permissions:(RLMSyncRealmPermission)permissions
+                               token:(RLMNotificationToken * _Nullable __autoreleasing * _Nonnull)token
+                            callback:(RLMSyncOfferChangeBlock)callback
+NS_REFINED_FOR_SWIFT;
+
+// TODO: does this really need a callback? Won't the object just be deleted immediately?
+/**
+ Revoke a permission offer which was previously generated.
+
+ @param offerKey        The identifier of the previously-generated permission
+                        offer (not the token).
+ @param callback        An optional callback block, to asynchronously report
+                        the progress of the operation.
+ @return    If a callback was specified, this will be a Realm notification
+            token. Call `-stop` if you want to deactivate the callback.
+ */
+- (nullable RLMNotificationToken *)deleteOffer:(NSString *)offerKey
+                                      callback:(nullable RLMSyncOfferChangeBlock)callback
+NS_REFINED_FOR_SWIFT;
+
+/**
+ Given a key representing an existing permission offer, register a callback
+ on that offer.
+
+ This method is useful if you wish to observe offers across app restarts.
+ The offer key can be persisted in whatever way is convenient (for example,
+ in a Realm), and then retrieved the next time the app is launched.
+
+ @param offerKey        The identifier of the previously-generated permission
+                        offer (not the token).
+ @param callback        A callback block, to asynchronously report
+                        the progress of the operation.
+ @return    A Realm notification token. Call `-stop` if you want to deactivate
+            the callback.
+ */
+- (nullable RLMNotificationToken *)addCallbackToExistingOffer:(NSString *)offerKey
```

Comment:
Given that we are a offline-first database, it seems natural to me to have this behavior as a part of the convenience API, but instead of having a separate method for it. Would it make sense to overload the other methods with one that just accept the token and the callback?